packages/dds/tree/src/forest/cursor.ts

```diff
@@ -4,89 +4,183 @@
  */
 
 import { assert } from "@fluidframework/common-utils";
-import { FieldKey, TreeType, Value } from "../tree";
+import { FieldKey, TreeType, UpPath, Value } from "../tree";
 
-export const enum TreeNavigationResult {
-    /** Attempt to navigate cursor to a key or index that is outside the client's view. */
-    NotFound = -1,
+/**
+ * A stateful low-level interface for reading tree data.
+ */
+export interface ITreeCursor {
+    /**
+     * What kind of place the cursor is at.
+     * Determines which operations are allowed.
+     */
 
-    /** Attempt to navigate cursor to a portion of the tree that has not yet been loaded. */
-    Pending = 0,
+    readonly mode: CursorLocationType;
+    /*
+     * True iff the current field or node (depending on mode) is "pending",
+     * meaning that is has not been downloaded.
+     */
+    readonly pending: boolean;
 
-    /** ITreeReader successfully navigated to the desired node. */
-    Ok = 1,
-}
+    /**
+     * Moves the "current field" forward one in an arbitrary field traversal order.
+     *
+     * If there is no remaining field to iterate to,
+     * returns false and navigates up to the parent setting the mode to `Nodes`.
+     *
+     * Order of fields only applies to this function,
+     * and is only guaranteed to be consistent thorough a single iteration.
+     *
+     * If mode is `Nodes` enters the first field.
+     *
+     * If skipPending, skip past fields which are currently pending.
+     * This can be used to skip to the end of a large number of consecutive pending fields.
+     *
+     * NOT allowed if mode is `Nodes` and also `pending`.
+     */
+    nextField(skipPending: boolean): boolean;
 
-/**
- * TreeNavigationResult, but never "Pending".
- * Can be used when data is never pending.
- */
-export type SynchronousNavigationResult = TreeNavigationResult.Ok | TreeNavigationResult.NotFound;
+    /**
+     * Moves `offset` nodes in the field.
+     * If seeking to exactly past the either end,
+     * returns false and navigates up to the parent field (setting `inFields` to `true`).
+     *
+     * If mode is `Fields`, enters "outside" the field range, then seeks `offset`.
```

Comment:
I don't understand what "outside" means in this context.

Reply:
I have rephrased it.